server/src/main/java/org/elasticsearch/cluster/routing/RelocationFailureInfo.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+
+/**
+ * Holds additional information as to why the shard failed to relocate.
+ */
+public record RelocationFailureInfo(int failedRelocations) implements ToXContentFragment, Writeable {
+
+    public RelocationFailureInfo {
+        assert failedRelocations > 0 : "Expect positive failures count, got: " + failedRelocations;
```

Comment:
I'm not a big fan of using `null` to represent `RelocationFailureInfo(0)`. Let's make the field non-nullable and use a singleton for `RelocationFailureInfo(0)` to avoid the unnecessary allocation & heap usage.

```suggestion
        assert failedRelocations >= 0 : "Expect non-negative failures count, got: " + failedRelocations;
```